docs/dev/hacking-guide.md

```diff
@@ -0,0 +1,233 @@
+
+If it is a cloud provider related change, you should fork and clone one of provider repositories, the list can be found above.
+
+If the change is related to all providers, you should fork and clone machine-api-operator.
+
+In order to add changes to our e2e test suite fork and clone [cluster-api-actuator-pkg](https://github.com/openshift/cluster-api-actuator-pkg).
+
+## How to run unit tests
+You may want to run unit tests before pushing changes. It can be done in a similar way for  Machine API Operator and cloud providers repositories.
+
+Prerequisite:
+```
+git checkout github.com/openshit/$repository_name
+cd $repository_name
+```
+In order to run the unit tests locally on your machine run the following command:
+```
+NO_DOCKER=1 make test
+```
+*Note* :
+If you run this command inside the machine-api-operator directory, it will run unit tests for machine, machineset, machine health check controllers and vsphere provider.
+If this command is run inside a cloud provider repository you will run only cloud provider specific tests.
+
+## How to run a component locally for testing
+### Running machine controller
+Prerequisites:
+```
+git checkout github.com/openshit/$repository_name
+cd $repository_name
+```
+Make sure your $KUBECONFIG is set properly, because it will be used to interact with your cluster.
+
+First step is to scale down the cluster version operator.
+
+“At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.” - https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html (Thanks for this post, Mike!)
+
+```
+oc scale --replicas=0 deploy/cluster-version-operator -n openshift-cluster-version
+```
+
+Second step is scaling down the machine-api-operator. The operator watches for machine controller to be running and will prevent it from scaling down.
+
+``` 
+oc scale deployment/machine-api-operator -n openshift-machine-api --replicas=0
+```
+
+Third step, once MAO is down scaled it is required to remove the machine-controller container from machine-api-controllers Deployment.
+
+```
+oc edit deployment/machine-api-controllers -n openshift-machine-api
+```
+
+Finally, once all has been scaled down you can compile and run the controller.
+
+```
+NO_DOCKER=1 make build
+ ./bin/manager -v 5
+```
+
+*Notes*:
+NO_DOCKER=1 will build the controller on your local machine and outside of any containers.
+The commands and binary names might slightly differ across providers
+
+## How to build the software in a container for remote testing
+Mike wrote a great article about this topic https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html, this section will be a copy-paste + some additional details.
```

Comment:
we can probably remove this sentence, i don't mind copying the material from my blog. i'm happy if we could just have a link back to the source as a reference.